### ISSUE TITLE
Non-zero status code returned while running Slice node. Name:'Slice_24' Status Message: slice.cc:153 FillVectorsFromInput Starts must be a 1-D array

### ISSUE THREAD

**Issue by lzhin** [microsoft/onnxruntime#11257]
**the code I export the onnx model in pytorch  is as follow:**

    offset = 0
    if args['left_chunks'] > 0:  # 16/4
        required_cache_size = args['chunk_size'] * args['left_chunks']
        offset = required_cache_size
        # Real cache
        att_cache = torch.zeros(
            (args['num_blocks'], args['head'], required_cache_size,
             args['output_size'] // args['head'] * 2))
        # Real mask
        att_mask = torch.ones(
            (args['batch'], 1, required_cache_size + args['chunk_size']),
            dtype=torch.bool)
        att_mask[:, :, :required_cache_size] = 0
    elif args['left_chunks'] <= 0:  # 16/-1, -1/-1, 16/0
        required_cache_size = -1 if args['left_chunks'] < 0 else 0
        # Fake cache
        att_cache = torch.zeros(
            (args['num_blocks'], args['head'], 0,
             args['output_size'] // args['head'] * 2))
        # Fake mask
        att_mask = torch.ones((0, 0, 0), dtype=torch.bool)
    cnn_cache = torch.zeros(
        (args['num_blocks'], args['batch'],
         args['output_size'], args['cnn_module_kernel'] - 1))
    inputs = (chunk, offset, required_cache_size,
              att_cache, cnn_cache, att_mask)
    torch.onnx.export(
        encoder, inputs, encoder_outpath, opset_version=14,
        export_params=True, do_constant_folding=True,
        input_names=[
            'chunk', 'offset', 'required_cache_size',
            'att_cache', 'cnn_cache', 'att_mask'
        ],
        output_names=['output', 'r_att_cache', 'r_cnn_cache'],
        dynamic_axes=dynamic_axes, verbose=False)`

**and  the  code running in c++ is:**

 int num_frames = cached_feature_.size() + chunk_feats.size();
  const int feature_dim = chunk_feats[0].size();
  int requried_cache_size = chunk_size_ * num_left_chunks_;
  std::vector<float> model_input;
    for (size_t i = 0; i < cached_feature_.size(); ++i) {
        for (int j = 0; j < feature_dim; j++) {
        model_input.emplace_back(cached_feature_[i][j]);
        }
    }
    for (size_t i = 0; i < chunk_feats.size(); ++i) {
        for (int j = 0; j < feature_dim; j++) {
        model_input.emplace_back(chunk_feats[i][j]);
        }
    }
  const int64_t input_shape[3] = {1, num_frames, feature_dim};
  
  std::vector<int64_t> offset{offset_};
  const int64_t offset_shape[1] = {1};
  
  std::vector<int64_t> requried_cache_size_vec{requried_cache_size};
  const int64_t requried_cache_size_shape[1] = {1};

    Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
        memory_info_, model_input.data(), model_input.size(), input_shape, 3);
    Ort::Value offset_tensor = Ort::Value::CreateTensor<int64_t>(
        memory_info_, offset.data(), 1, offset_shape, 1);

    
    Ort::Value requried_cache_size_tensor = Ort::Value::CreateTensor<int64_t>(
        memory_info_, requried_cache_size_vec.data(), 1,
        requried_cache_size_shape, 1);
    std::vector<Ort::Value> tensors;
    tensors.emplace_back(std::move(input_tensor));
    tensors.emplace_back(std::move(offset_tensor));
    tensors.emplace_back(std::move(requried_cache_size_tensor));
    tensors.emplace_back(std::move(att_cache_ort_));
    tensors.emplace_back(std::move(cnn_cache_ort_));
    tensors.emplace_back(std::move(att_mask_ort_));

    std::vector<Ort::Value> ort_outputs =
        encoder_session_->Run(Ort::RunOptions{nullptr}, input_names_,
                                       tensors.data(), 6, output_names_, 3);

**when run this code, the error is:**

Non-zero status code returned while running Slice node. Name:'Slice_24' Status Message: slice.cc:153 FillVectorsFromInput Starts must be a 1-D array

**I would to know how to create the offset and requried_cache_size  variable**

**Comment by pranavsharma** [comment]
Can you attach the onnx model?

**Comment by lzhin** [comment]
> Can you attach the onnx model?

ok, this is the onnx model url :  https://athena-genieopen-gz-1253539814.cos.ap-guangzhou.myqcloud.com/athena-wenet/encoder.onnx

**Comment by Mddct** [comment]
<img width="855" alt="Screen Shot 2022-04-24 at 12 18 53" src="https://user-images.githubusercontent.com/4906435/164956443-8a048b3c-90c8-4ae4-ac38-2c623e2e2926.png">
offset is no a 1-d tensor in export_onnx ,  but use a "1-d array" tensor" in your c++ code , an than use a "unsequence" the slice input shape is 2-d tensor

**Comment by lzhin** [comment]
> 

this is the code that slice by offset :
`
        assert offset + x.size(1) < self.max_len
        self.pe = self.pe.to(x.device)
        pos_emb = self.pe[:, offset:offset + x.size(1)]
        x = x * self.xscale + pos_emb
`
I don't find the unsqueeze operation.

The fucntion to export the onnx model is :
`
    def forward_chunk(
        self,
        xs: torch.Tensor,
        offset: int,
        required_cache_size: int,
        att_cache: torch.Tensor = torch.zeros(0, 0, 0, 0),
        cnn_cache: torch.Tensor = torch.zeros(0, 0, 0, 0),
        att_mask: torch.Tensor = torch.ones((0, 0, 0), dtype=torch.bool),
    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
        assert xs.size(0) == 1
        # tmp_masks is just for interface compatibility
        tmp_masks = torch.ones(1,
                               xs.size(1),
                               device=xs.device,
                               dtype=torch.bool)
        tmp_masks = tmp_masks.unsqueeze(1)
        if self.global_cmvn is not None:
            xs = self.global_cmvn(xs)
        # NOTE(xcsong): Before embed, shape(xs) is (b=1, time, mel-dim)
        xs, pos_emb, _ = self.embed(xs, tmp_masks, offset)
............
`

I know the parameter offset should not be create by createtensor function as the offset is int in python.
But I don't know what fucntion to create the offset value in c++ runtime.
Could you give me some advices about how to create the offset value? Thanks.

**Comment by Mddct** [comment]
firstly torch slice will be converted to unsequeze and slice operation。

secondly，create a tensor with a empty shape  like this

int64_t offset = ……

Ort::Value offset_tensor = Ort::Value::CreateTensor<int64_t>( memory_info_, &offset, 1, std::vector<int>{}.data()，0);

**Comment by lzhin** [comment]
> firstly torch slice will be converted to unsequeze and slice operation。
> 
> secondly，create a tensor with a empty shape like this
> 
> int64_t offset = ……
> 
> Ort::Value offset_tensor = Ort::Value::CreateTensor<int64_t>( memory_info_, &offset, 1, std::vector{}.data()，0);

**Thanks for your advices. Now the slice error has been eliminated.
But the another error appears:**
2022-04-24 20:07:18.419115816 [E:onnxruntime:, sequential_executor.cc:346 Execute] Non-zero status code returned while running Add node. Name:'Add_99' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/math/element_wise_ops.h:583 onnxruntime::Broadcaster::Broadcaster(const std::vector<long int>&, const std::vector<long int>&) largest <= 1 was false. Can broadcast 0 by 0 or 1. 64 is invalid.

terminate called after throwing an instance of 'Ort::Exception'
  what():  Non-zero status code returned while running Add node. Name:'Add_99' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/math/element_wise_ops.h:583 onnxruntime::Broadcaster::Broadcaster(const std::vector<long int>&, const std::vector<long int>&) largest <= 1 was false. Can broadcast 0 by 0 or 1. 64 is invalid.

I don't know how to deal with it. Do you know what's the problem?

**Comment by Mddct** [comment]
you may need to rewrite required cache size tensor in same way

**Comment by lzhin** [comment]
> you may need to rewrite required cache size tensor in same way

**I have changed the offset and requried_cache_size before. And the above error occurred.**

  Ort::Value offset_tensor = Ort::Value::CreateTensor<int64_t>(
      memory_info_, &offset_, 1,  std::vector<int64_t>{}.data(), 0);

  Ort::Value requried_cache_size_tensor = Ort::Value::CreateTensor<int64_t>(
      memory_info_, &requried_cache_size, 1,
      std::vector<int64_t>{}.data(), 0);

**Comment by Mddct** [comment]
offset= ? cache size =？ maybe you have a wrong value and cause you have a bad position embedding

**Comment by lzhin** [comment]
> offset= ? cache size =？ maybe you have a wrong value and cause you have a bad position embedding

**this is the encoder forword:**

void OnnxAsrModel::ForwardEncoderFunc(
    const std::vector<std::vector<float>>& chunk_feats,
    std::vector<std::vector<float>>* out_prob) {
  int num_frames = cached_feature_.size() + chunk_feats.size();
  const int feature_dim = chunk_feats[0].size();
  int requried_cache_size = chunk_size_ * num_left_chunks_;
  std::vector<float> model_input;
  for (size_t i = 0; i < cached_feature_.size(); ++i) {
    for (int j = 0; j < feature_dim; j++) {
      model_input.emplace_back(cached_feature_[i][j]);
    }
  }
  for (size_t i = 0; i < chunk_feats.size(); ++i) {
    for (int j = 0; j < feature_dim; j++) {
      model_input.emplace_back(chunk_feats[i][j]);
    }
  }
  const int64_t input_shape[3] = {1, num_frames, feature_dim};

  // std::vector<int64_t> offset{offset_};
  const int64_t offset_shape[0];
  int64_t offset=offset_;

  std::vector<int64_t> requried_cache_size_vec{requried_cache_size};
  const int64_t requried_cache_size_shape[1] = {1};

  Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
      memory_info_, model_input.data(), model_input.size(), input_shape, 3);
  Ort::Value offset_tensor = Ort::Value::CreateTensor<int64_t>(
      memory_info_, &offset_, 1, offset_shape, 0);

  Ort::Value requried_cache_size_tensor = Ort::Value::CreateTensor<int64_t>(
      memory_info_, requried_cache_size_vec.data(), 1,
      requried_cache_size_shape, 1);

  std::vector<Ort::Value> tensors;
  tensors.emplace_back(std::move(input_tensor));
  tensors.emplace_back(std::move(offset_tensor));
  tensors.emplace_back(std::move(requried_cache_size_tensor));
  tensors.emplace_back(std::move(att_cache_ort_));
  tensors.emplace_back(std::move(cnn_cache_ort_));
  tensors.emplace_back(std::move(att_mask_ort_));

  std::vector<Ort::Value> ort_outputs =
      encoder_session_->Run(Ort::RunOptions{nullptr}, input_names_,
                            tensors.data(), 6, output_names_, 3);

  offset_ += ort_outputs[0].GetTensorTypeAndShapeInfo().GetShape()[1];
  att_cache_ort_ = std::move(ort_outputs[1]);
  cnn_cache_ort_ = std::move(ort_outputs[2]);

  std::vector<Ort::Value> ctc_tensors;
  ctc_tensors.emplace_back(std::move(ort_outputs[0]));

  std::vector<Ort::Value> ctc_ort_outputs =
      ctc_session_->Run(Ort::RunOptions{nullptr}, ctc_input_names_,
                        ctc_tensors.data(), 1, ctc_output_names_, 1);
  encoder_outs_.push_back(std::move(ctc_tensors[0]));

  float* logp_data = ctc_ort_outputs[0].GetTensorMutableData<float>();
  auto type_info = ctc_ort_outputs[0].GetTensorTypeAndShapeInfo();

  int num_outputs = type_info.GetShape()[1];
  int output_dim = type_info.GetShape()[2];
  out_prob->resize(num_outputs);
  for (int i = 0; i < num_outputs; i++) {
    (*out_prob)[i].resize(output_dim);
    memcpy((*out_prob)[i].data(), logp_data + i * output_dim,
           sizeof(float) * output_dim);
  }
}


**I'll check the code carefully tomorrow.
I have submited a draft pr in [wenet pr](https://github.com/wenet-e2e/wenet/pull/1053). Look forword your review. Thank you very much.**

**Comment by Ryuk17** [comment]
Have you solved this problem ? I meet the save problem when testing the onnx model transferred from torch in WeNet.

**Comment by lzhin** [comment]
> Have you solved this problem ? I meet the save problem when testing the onnx model transferred from torch in WeNet.

Onnx export and inference have been implemented in wenet. https://github.com/wenet-e2e/wenet/blob/main/runtime/core/decoder/onnx_asr_model.h
